### PR TITLE
Add page on creating custom widgets to docs

### PR DESCRIPTION
**Pull request by elParaguayo**
Here's a first draft of documentation on how to create a widget. I hope it's of use/interest!

You may have seen the discussion (#2138) about where's the best place for documentation like this. My feeling is that it should be in the official docs (you'll see I've created a section called `Tips & Tricks`) to separate it from the more formal section.

This is a draft request for now as it refers to my dbus_next PR #1994 for instructions on how widgets can use dbus.

### CODE REVIEW

**Comment by elParaguayo** [comment]
@ramnes  @tych0 Just curious: is this something that could be included in the main docs?

**Comment by tych0** [comment]
I don't see why not, although I haven't look closely since you marked it a draft.

**Comment by m-col** [comment]
I just read through it, it's brilliant and very thorough, definitely something we should have!

**Comment by elParaguayo** [comment]
> I don't see why not, although I haven't look closely since you marked it a draft.

That's fine - just wanted to check that I wasn't wasting my time on this.


> I just read through it, it's brilliant and very thorough, definitely something we should have!

Thank you. Will work on more sections then.

**Comment by elParaguayo** [comment]
This should be good to add now that #1994 is merged.

**Comment by m-col** [comment]
LGTM. Best to get a second pair of eyes over it though :)

**Review by m-col — APPROVED** [review]

File: docs/manual/howto/widget.rst [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+
+The __init__ method
+-------------------
+
+Parameters that should not be changed by users can be defined in the ``__init__``
+method.
+
+This method is run when the widgets are initially created. This happens before
+the ``qtile`` object is available. This section is therefore appropriate for
+hard coding values.
```
Comment (tych0) [inline docs/manual/howto/widget.rst:171]: I don't really understand what this last sentence means.

File: docs/manual/howto/widget.rst [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+Once this method has been run, your widget should be ready to display content as
+the bar will draw once it has finished its configuration.
+
+Calls to methods required to prepare the content for your widget should therefore
+be made from this method rather than ``__init__``.
+
+Displaying output
+=================
+
+As noted at the beginning of this page, a Qtile widget is just a drawing that is
```
Comment (tych0) [inline docs/manual/howto/widget.rst:189]: I think you can drop the first bits before the comma.

**Comment by tych0** [comment]
Two minor nits, but overall this looks great!

File: docs/manual/howto/widget.rst (hunk 1 above)
Comment (elParaguayo) [reply docs/manual/howto/widget.rst:171]: This is in the context of where to define parameters/attributes but, you're right, it doesn't add anything so will remove it.

File: docs/manual/howto/widget.rst (hunk 2 above)
Comment (elParaguayo) [reply docs/manual/howto/widget.rst:189]: Agreed.

**Pull request merged**